### PR TITLE
Fix as proposed by Johns Hopkins researchers

### PR DESCRIPTION
**Pull request by tomharrisonjr**
https://app.asana.com/0/1199236100187165/1201654099877592/f

```
This message is to inform you of a vulnerability that exists in the JavaScript served from spreedly.com. A team of researchers from the Johns Hopkins University (JHU) are working on a tool that detects a new type of vulnerability in client-side JavaScript called Prototype Pollution Vulnerabilities. These vulnerabilities can permit a malicious actor to inject a property with an arbitrary value to a prototypical object, such as Object.prototype, and then launch attacks such as Cross-site Scripting (XSS) and cookie/URL manipulations. Then, if XSS is present, attackers can use malicious code thus injected to collect any information your site stores in JavaScript objects on the client; if cookie/URL manipulation is present, attackers can launch session fixation or Cross-site Request Forgery (CSRF).

The details of the vulnerability specific to your website as well as a recommended fix are:

Vulnerability category: client-side prototype pollution.
Potential consequences: Cross-Site Scripting, Denial-of-Service, Remote Code Execution, Session Hijacking, etc.
How to trigger: When we navigate to https://www.spreedly.com/professional-services/?__proto__[123]=testv&__proto__.123=testv&constructor[prototype][123]=testv , the JavaScript code sets the string "Value" into Object.prototype["123"] . This would cause a client-side prototype pollution vulnerability. The adversary can set "123" and "Value" to arbitrary strings, persuade the victim to visit this crafted URL through social engineering and launch an attack such as Cross-Site Scripting, Denial-of-Service, Remote Code Execution and Session Hijacking.
Vulnerable code location: In the JavaScript/HTML file https://x.clearbitjs.com/v2/pk_b537c18b2ca7409f52cdd3b559bdba0a/tracking.min.js. The line number is 4. The function name is parse. The vulnerable codes start with function (str){if("string"!=typeof str)return{};str=trim(str);if(""==str)return{};if("?"==str.charAt(0))str=str.slice(1);var obj={};var pairs=str.split("&");for(var i=0;i<pairs.length;i++){var parts=pairs[i].split("=");var key=decode(parts[0]);var m;if(m=pattern.exec(key)){obj[m[1]]=obj[m[1]]||[];obj[m[1]][m...

.
How to patch: Add a function that can encode the strings "__proto__", "constructor", "prototype":
sanitizeKey = function(t) {

return t && ["__proto__", "constructor", "prototype"].includes(t.toLowerCase()) ? t.toUpperCase() : t

}

and apply that function before setting any property from the URL-search query strings. For instance, replace each “obj[property]” with “obj[sanitizeKey(property)]” within that function.

How Our Tool Found the Vulnerability:

The tool we have developed issues an HTTP GET request to the domain and analyzes the JavaScript that the web page contains or links to. This analysis is non-invasive and analyses the code that your server sends every time this page is requested. Following automated analysis, our researchers entered benign code via the injection method described above. IT IS IMPORTANT TO NOTE that this injection occurred only on our systems as a proof of concept for how a malicious actor would exploit this vulnerability. Since this is a client-side vulnerability, no interaction beyond the initial HTTP GET request occurred with your servers.

Disclosure Timetable:

JHU follows a process of responsible disclosure. Following disclosure of vulnerabilities, we adhere to a non-disclosure period. During this time, we do not publicly disclose the vulnerabilities, however we may continue to research them and submit academic papers on the matter. Our tool is the subject of a paper to appear in the Proceedings of Network & Distributed System Security Symposium (NDSS), 2022. The paper will be available with broader distribution at the conference in February 27th, 2022.

If you have any further questions, do not hesitate to reach out to any of us at the email addresses below. Finally, we would appreciate an email in reply acknowledging receipt of this disclosure.

Regards,

Zifeng Kang (Ph.D. Student, Computer Science Department, Johns Hopkins University, email: zkang7@jhu.edu)

Song Li (Ph.D. Candidate, Computer Science Department, Johns Hopkins University, email: lsong18@jhu.edu)

Yinzhi Cao (Assistant Professor, Computer Science Department, Johns Hopkins University, email: yzcao@cs.jhu.edu)

Joseph Carrigan (Senior Security Engineer, Vulnerability Disclosure Coordinator, Johns Hopkins University, Information Security Institute, email: joseph.carrigan@jhu.edu)
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -61,6 +61,8 @@ exports.parse = function(str){
     var parts = pairs[i].split('=');
     var key = decode(parts[0]);
     var m;
+    // Sanitize keys to uppercase to mitigate client-side prototype pollution
+    key && ["__proto__", "constructor", "prototype"].includes(key.toLowerCase()) ? key.toUpperCase() : key
```
Comment (harlow) [inline index.js]: should we be setting `key` as a receiver here? I don't think the functions augment the variable.

```
key = 'hello'
'hello'
key.toUpperCase()
'HELLO'
key
'hello'
```

maybe something like this?

```javascript
if key && ["__proto__", "constructor", "prototype"].includes(key.toLowerCase()) {
 key = key.toUpperCase() 
}
```

Comment (davidlumley) [reply index.js]: Great catch — `toUpperCase()` returns a value, doesn't mutate in place.

**Review by davidlumley — APPROVED** [review]
Nice patch Tom!

File: index.js (hunk 1 above)
Comment (tomharrisonjr) [reply index.js]: Thanks!  Implemented ^^

**Pull request merged**